Add `deltaSum` aggregate function, docs & test

rf opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

This PR adds `deltaSum` aggregation function, which sums differences between consecutive rows.

It is similar to `runningDifference` function, but is implemented as an aggregate function.

This is useful for IOT metrics where we have a counter that continuously increments until it hits some max value, then is reset to 0 or some other small value.

Please let me know if there's anything else I can do to get this pulled in -- we need this in order to adopt Clickhouse for our business.

Thanks!


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

...
Add aggregate function `deltaSum` for summing the differences between consecutive rows.

Detailed description / Documentation draft:

...

rf committed:
Add `deltaSum` aggregate function, docs & test

rf committed:
record seen first and last

rf committed:
fix tests

rf committed:
correct conditional

rf committed:
clang-format

rf committed:
correct docs

rf committed:
correct merge logic

rf committed:
merge tests

rf committed:
check arg type

rf committed:
turn off ubsan for deltasum functions

kitaisreal commented on src/AggregateFunctions/AggregateFunctionDeltaSum.h:
```diff
@@ -0,0 +1,108 @@
+        }
+        else
+        {
+            this->data(place).sum += this->data(rhs).sum;
```
Consider to use reference to this->data(place) and add some comments for each branch.

kitaisreal commented on src/AggregateFunctions/AggregateFunctionDeltaSum.h:
```diff
@@ -0,0 +1,108 @@
+    T last = 0;
+    bool seen_first = false;
+    T first = 0;
+};
```
Consider to rearrange values in this structure like:
```
template <typename T>
struct AggregationFunctionDeltaSumData
{
     T sum = 0;
     T last = 0;
     T first = 0;
     bool seen_first = false;
     bool seen_last = false;
};
```
Example:

```
template <typename T>
struct First
{
     T sum = 0;
     bool seen_last = false;
     T last = 0;
     bool seen_first = false;
     T first = 0;
};

template <typename T>
struct Second
{
     T sum = 0;
     T last = 0;
     T first = 0;
     bool seen_last = false;
     bool seen_first = false;
};
```

If for example T will be uint64_t sizeof(First) == 40, but sizeof(Second) == 32

kitaisreal commented on src/AggregateFunctions/AggregateFunctionDeltaSum.cpp:
```diff
@@ -0,0 +1,49 @@
+
+    DataTypePtr data_type = arguments[0];
+
+    if (!isNumber(data_type))
```
As I see is number definition it can support decimal.
```
template <typename T>
inline bool isNumber(const T & data_type)
{
    WhichDataType which(data_type);
    return which.isInt() || which.isUInt() || which.isFloat() || which.isDecimal();
}
```
But in implementation we assume that argument column is numeric without decimal:
```
auto value = static_cast<const ColumnVector<T> &>(*columns[0]).getData()[row_num];
```
For decimal there is ColumnDecimal.

kitaisreal commented on tests/queries/0_stateless/01700_deltasum.sql:
```diff
@@ -0,0 +1,7 @@
+select deltaSum(arrayJoin([1, 2, 3]));
```
It seems like this function support floats, decimals as well consider adding some test cases for them.

rf committed:
Merge remote-tracking branch 'origin/master' into rf/add-deltasum

rf committed:
feedback: use references, dont support decimal, rearrange struct members

rf committed:
dont allow col type to be decimal

rf commented:
@kitaisreal thanks so much for the feedback! I updated the branch and resolved the addressed comments.

I decided to not add support for `Decimal` type for now as we don't need it. If it's necessary to add support for `Decimal` to get this change in I'll figure it out.

rf committed:
style fixes

kitaisreal committed:
Update deltasum.md
